Fix fatal error auto-loading file for already existing class by invalid class-name

ktomk opened the pull request:
... filing fix for #5024

- [x] Regression test for invalid class-names
- [x] Give Phpstan a helping hand on anonymous class in test-suite
- [x] Add the fix (ltrim)

Using invalid class-names, specifically with two consecutive backslashes, while performing `class_exist()` checks, triggers auto-loading for the invalid class-name which then may result in the re-inclusion of existing files (due to the mapping of the class-name to a file-system-path) which then may trigger a fatal error if such a file contains the class-definition of an already defined
class.

          load class: Name\Space\DoubleTrouble
    file-system path: ../Name/Space/DoubleTrouble.php

    class_exists(Name\Space\\DoubleTrouble)
                           ^^
       file-system path: ../Name/Space//DoubleTrouble.php
                                      ^^
    already loaded file: ../Name/Space/DoubleTrouble.php

This fatal error brings rector to halt.

To stabilize class_exist() while auto-loading is active (it is always active w/ rector), prevent invalid class-names, specifically containing two or more consecutive backslashes.

Fix is to prevent consecutive backslashes when putting the name-space and the class-name together by dropping all leading backslashes from the class-name.

TomasVotruba commented on rules/renaming/tests/Rector/FileWithoutNamespace/PseudoNamespaceToNamespaceRector/Fixture/fixture6.php.inc:
```diff
@@ -12,6 +12,13 @@ class Fixture6
+     */
+    public function testMore(\DateTime $time)
+    {
+    }
```
Could you decoupe this to a new test fixture file? Just so the case is isolated

ktomk replied:
Sure, I put it into a new fixture file, this one was just a quick shot to get this running on CI here.

TomasVotruba commented on rules/type-declaration/src/PHPStan/Type/ObjectTypeSpecifier.php:
```diff
@@ -128,7 +128,9 @@ private function matchSameNamespacedObjectType(Node $node, ObjectType $objectTyp
-        $namespacedObject = $namespaceName . '\\' . $objectType->getClassName();
+        // ltrim: prevents double-backslash that gives fatal error with existing class for class_exists() while
+        // auto-loading #5024
+        $namespacedObject = $namespaceName . '\\' . ltrim($objectType->getClassName(), '\\');
```
Does this fix the bug you had?

Could you run first CI without this fix, so we can see the reported bug? Thanks :+1:

ktomk replied:
Yes, this is identified as effective against the fatal error. I'll do the regression test first, this includes auto-loading.

ktomk replied:
CI without this fix is in: https://github.com/rectorphp/rector/pull/5026/commits/88163d6c5f6b315d9649392f2aeb34fa7f3424d7

Phpstan is exceptionally low on anonymous classes and over-estimates itself to not infer the anonymous type (Phpstorm does this). If you know by chance a docblock notation to handle this please let me know, also about suppressing these, I'm not as fluent with Phpstan as you are.

ktomk replied:
Noise reduced by suppression. The one is a known issue with Phpstan (https://github.com/phpstan/phpstan/issues/3144) and the other is an issue with Simplify (just reported: https://github.com/symplify/symplify/issues/2717).

TomasVotruba commented:
Not sure if that would help, but there is also `doTestFileInfoWithoutAutoload()` method, that will test the file without autoloading it.

ktomk committed:
Regression test class-name validity

In PseudoNamespaceToNamespaceRectorTest

For class_exists() which triggers the auto-loader, extend the test-case
to keep track of all auto-loaded classes and assert that all class-names
used in auto-loading are valid class-names.

NOTE: The class-names in the PseudoNamespaceToNamespaceRector are not
      meant to be actually existing class-names, but just some that
      might be and there-fore always valid ones.

The problem with invalid class-names is specifically, that they never
exist and trigger the auto-loader always. If they contain two consecutive
backslashes (name-space-separator) then the mapping to the file-system
might merge them to an already included file which will make rector exit
with a fatal error as a class is getting re-defined by the auto-load
include:

> Fatal error: Cannot declare class <class-name>, because the name is
> already in use in ...

Therefore the PseudoNamespaceToNamespaceRector must not trigger
auto-loading on invalid class-names.

Fixtures known to fail:

- fixture4.php.inc
- fixture5.php.inc
- fixture6.php.inc
- var_doc.php.inc

Notes on Phpstan supressions (others should be self-explanatory):

- getAutoloadClassNameAssertion() has no return-type hint by intention
  as static code-analyzers (Phpstorm) won't be able to infer the concrete
  type of the anonymous class returned.

- bogus error for anonymous classes:

        '#Class "AnonymousClassb3d71a09d1e123ac8b2eb26b93f1b476" should have suffix "TestCase" by parent class/interface#'

TomasVotruba commented on rules/renaming/tests/Rector/FileWithoutNamespace/PseudoNamespaceToNamespaceRector/PseudoNamespaceToNamespaceRectorTest.php:
```diff
@@ -42,4 +48,65 @@ protected function getRectorsWithConfiguration(): array
+
+            public function __destruct()
+            {
+                \spl_autoload_unregister($this);
```
This must be handled without hacking autoload, so it's easy to maintain in years to come

ktomk replied:
Yes sure this must be stable as the foundations of mountains. Therefore it would help me if you could elaborate a bit where this spl_autoload_register / spl_autoload_unregister scheme poses a maintenance burden? Actually I can not imagine that it will go away nor that composer drops it. @TomasVotruba

TomasVotruba replied:
Autoload is handled by composer, so we can autoload it by putting in psr-4 scope or not autoloadit by putting anywhere else. No need to create 2nd composer :)

ktomk replied:
Yes, sure. But this is all not happening with the change-set :) It just keeps a list of class-names to assert it later ...

TomasVotruba replied:
See https://github.com/rectorphp/rector/pull/5040/files

ktomk committed:
Fix invalid class-name concatenating w/ name-space

Previously ObjectTypeSpecifier::matchSameNamespacedObjectType() was prone
to create invalid class-names with two consecutive backslashes.

These can lead to fatal errors as class_exists() always returns false
for them while triggering an auto-load which may resolve to a file with
a class-definition that is already defined (as the class-name with a
single backslash would be valid and already loaded).

Rector then exists with a fatal error.

> Fatal error: Cannot declare class <class-name>, because the name is
> already in use in ...

Fix is to prevent this fatal error cause by auto-loading by preventing
consecutive backslashes when concatenating the namespace-object class-name
string.

Either the class is already loaded then class_exists() does not trigger the
auto-loading any longer -or- it does trigger the auto-loading and it can
eventually be loaded and would never clash.

Ref-Issue: #5024

TomasVotruba commented:
I'm looking into this and I can't find the test fixture in `/Fixtures` directory.
Where is it located?

TomasVotruba commented:
I'm running https://github.com/ktomk/rector-report now, I see the bug I look for, so all good for now :+1:

ktomk commented:
@TomasVotruba: Yes, I also put the bugfix-branch as a job on rector-report so this extra test is visible.

TomasVotruba commented:
Manaully merged in https://github.com/rectorphp/rector/pull/5040

Thank you :+1:

ktomk commented:
Huh, why did you merge changes that weren't accepted? Are you that much in a rush?

And thanks for having me.

TomasVotruba commented:
I merged only the accepted changes :) 

This was an edge case that required local fixture of `DateTime.php`, which is not allowed. It needed to excluded from typical fixtures, no `.php.inc`
